File: go/0496.go
package leetcode

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	ans := make([]int, len(nums1))
	indexOf := make(map[int]int, len(nums1))
	searchList := make([]int, 0)

	for index, num := range nums1 {
		indexOf[num] = index
	}

	for _, num := range nums2 {

		for i := 0; i < len(searchList); i++ {
			toBeat := searchList[i]
			if num > toBeat {
				ans[indexOf[toBeat]] = num
				searchList[i] = searchList[len(searchList)-1]
				searchList = searchList[:len(searchList)-1]
				i--
			}
		}

		_, ok := indexOf[num]

		if ok {
			searchList = append(searchList, num)
		}
	}

	for _, num := range searchList {
		ans[indexOf[num]] = -1
	}

	return ans
}

File: java/0496.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n = nums1.length;
        int m = nums2.length;
        int[] ans = new int[n];
        Map<Integer, Integer> nums2IndexHash = new HashMap<Integer, Integer>();

        for (int i = 0; i < m; i++) {
            nums2IndexHash.put(nums2[i], i);
        }

        int ind;
        for (int i = 0; i < n; i++) {
            ind = nums2IndexHash.get(nums1[i]);
            while (ind < m) {
                if (nums2[ind] > nums1[i]) {
                    ans[i] = nums2[ind];
                    break;
                }
                ind++;
            }
            if (ind == m)
                ans[i] = -1;
        }

        return ans;
    }
}